Core/include/Acts/Seeding/Seedfinder.ipp

```diff
@@ -32,268 +34,520 @@ Seedfinder<external_spacepoint_t>::Seedfinder(
-      if (deltaR < m_config.deltaRMin) {
-        continue;
-      }
-      if (deltaR > m_config.deltaRMax) {
-        break;
-      }
-
-      float cotTheta = (topSP->z() - zM) / deltaR;
-      if (std::fabs(cotTheta) > m_config.cotThetaMax) {
-        continue;
-      }
-      float zOrigin = zM - rM * cotTheta;
-      if (zOrigin < m_config.collisionRegionMin ||
-          zOrigin > m_config.collisionRegionMax) {
-        continue;
-      }
-      compatTopSP.push_back(topSP);
-    }
+    // no top SP found -> try next spM
     if (compatTopSP.empty()) {
       continue;
     }
+
+    auto end_DS = std::chrono::system_clock::now();
+    std::chrono::duration<double> elapse_DS = end_DS-start_DS;
+    std::get<0>(t_metric) += elapse_DS.count();
+    
     // contains parameters required to calculate circle with linear equation
+
+    auto start_TC = std::chrono::system_clock::now();
+    
     // ...for bottom-middle
     std::vector<LinCircle> linCircleBottom;
     // ...for middle-top
     std::vector<LinCircle> linCircleTop;
-    transformCoordinates(compatBottomSP, *spM, true, linCircleBottom);
-    transformCoordinates(compatTopSP, *spM, false, linCircleTop);
+    
+    SeedfinderCpuFunctions<external_spacepoint_t,sp_range_t>::transformCoordinates(compatBottomSP, *spM, true, linCircleBottom);
+    SeedfinderCpuFunctions<external_spacepoint_t,sp_range_t>::transformCoordinates(compatTopSP, *spM, false, linCircleTop);
 
-    // create vectors here to avoid reallocation in each loop
-    std::vector<const InternalSpacePoint<external_spacepoint_t>*> topSpVec;
-    std::vector<float> curvatures;
-    std::vector<float> impactParameters;
+    auto end_TC = std::chrono::system_clock::now();
+    std::chrono::duration<double> elapse_TC = end_TC-start_TC;
+    std::get<1>(t_metric) += elapse_TC.count();
+
+    auto start_TS = std::chrono::system_clock::now();
+    
+    auto seedsPerSpM = SeedfinderCpuFunctions<external_spacepoint_t,sp_range_t>::searchTriplet(*spM, compatBottomSP, compatTopSP, linCircleBottom, linCircleTop, m_config);
+    m_config.seedFilter->filterSeeds_1SpFixed(seedsPerSpM, outputVec);
+
+    auto end_TS = std::chrono::system_clock::now();
+    std::chrono::duration<double> elapse_TS = end_TS-start_TS;
+    std::get<2>(t_metric) += (elapse_TS).count();
+  }  
+
+  auto end_wall = std::chrono::system_clock::now();
+  std::chrono::duration<double> elapse_wall = end_wall-start_wall;
+  std::get<3>(t_metric) += elapse_wall.count();
+  
+  return outputVec;
+  }
+
+#ifdef  ACTS_HAS_CUDA
+  
+  // CUDA seed finding
+  template< typename external_spacepoint_t, typename platform_t>
+  template< typename T, typename sp_range_t>
+  typename std::enable_if< std::is_same<T, Acts::CUDA>::value, std::vector<Seed<external_spacepoint_t> > >::type
+  Seedfinder<external_spacepoint_t, platform_t>::createSeedsForGroup(
+    sp_range_t bottomSPs, sp_range_t middleSPs, sp_range_t topSPs) const{
+  std::vector<Seed<external_spacepoint_t>> outputVec;
```

Comment:
Ok so if I understand correctly, you use SFINAE to enable this if `platform_t` is `Acts::CUDA`. Maybe it would be possible to have either a class template or a function template sepcialization of that value, and then move and restrict the specialization to `Acts::CUDA` to a plugin/dedicated target. This way you should be able to get rid of all of the `#ifdef`s, as well as extracting the CUDA specific code into separate files, and cleanly move the CUDA aware code to only the plugin.

Reply:
Seconded, I don't think having both implementations in the same file is desirable. Also, instead of CPU_ImplementationClass::functionName() you could use Class<CPU>::functionName for consistency.

Reply:
@paulgessinger @robertlangenberg I just moved (almost) everything to the new plugin. Please review them